### DISCUSSION TITLE
More plugin events for guest actions

### DISCUSSION THREAD

**Discussion by spazzylemons** [OpenRCT2/OpenRCT2#18703] [Feature requests]
Adding more plugin events for guest actions, such as entering/leaving rides, purchasing items, and thoughts, can allow plugins to better work with guest information, such as tracking individual guests and connecting the information to external services, or in trainers or statistical projects to better analyze guest behavior and trends. I have started work on this myself at https://github.com/spazzylemons/OpenRCT2/ for personal use, but would like to create a discussion here to see others' thoughts and/or if other work has been done on this feature.

**Comment by IntelOrca** [comment]
The changes seem fine, we might want to improve some of the property and hook names. The addition of the thought list to `Guest` is something that should definitely be done.

**Reply by spazzylemons** [reply]
Sounds good, what improvements should be made?

**Reply by IntelOrca** [reply]
That can be decided when you create a pull request. Are you able to create a seperate one for the guest thoughts so that isn't in the same one as all the new hooks?

**Reply by spazzylemons** [reply]
Made PR #18732 for just guest thoughts

**Comment by dvasdekis** [comment]
I'm always disappointed by the existing AI (e.g. not riding transport rides to get around) and so I've had the dumb idea of using a Large Language Model (e.g. [Dalai Alpaca 7B](https://github.com/cocktailpeanut/dalai#2-request), already built in JS) to control the guests. My thoughts on flow would be:

1. convert current agent state to a free text question
2. pass to model
3. attempt to interpret answer back into RCT2

Future versions could start with a randomised personality for the agent, feed that plus the surrounds every time the agent does something to the model, and then have the model choose what to do next within the park. As a result of this, my hope is that you'd then get behaviour that's more "realistic".

So from what I can see, agent thoughts are now available, but I'd also like to pull back agent 'state' - e.g. what the agent can see in their local surrounds, to give the LLM more information to work with. Is there any API available for this, or one we could build that extends `guests`? I'm looking for some doco to help out with it!

**Reply by James103** [reply]
As far as I can tell, this may be much too resource intensive to be viable for parks with more than a couple of guests or staff.

With the language model being used, if it takes 1 second and 4 GB RAM for the model to run with 4 cores, each guest pings the model once per second on average during its AI routine (both very conservative) and each guest has their own instance of the model, then the hardware requirements get very extreme even for small parks, such that the hardware cost, in real-life money, exceeds the park construction cost in in-game money.
* One playthrough of "Forest Frontiers": 250 guests, 1,000 CPU cores and 1 **tera**byte of RAM
* A playthrough to completion of "Extreme Heights": 4,500 guests, over **18,000** CPU cores and 18 TB of RAM.

(Replace TB with GB in "X TB of RAM" if all the guests share one instance of the model)

**Reply by dvasdekis** [reply]
@James103 Thanks for the reply! You're absolutely correct about the situation today, but I think you're discounting some trends that could change things in the very near future.

1. These LLMs can now run on a consumer GPU and come in different sizes and requirements - see https://github.com/kuleshov/minillm and a very significant reduction in RAM requirements was obtained [here](https://github.com/ggerganov/llama.cpp/discussions/638?fbclid=IwAR2QkBNtxE0IicGcmplrFBUzPzwJN_xjTHIq5WdSrofsl4vrfc1eUTTz-nw#discussioncomment-5492916).
2. The very smallest LLM (probably 1/10th the size of llama-7b-4bit) is probably enough to drive quite an improvement in AI reasoning over the default RCT park visitor AI. It should be quite possible to find a tiny hobbled model that's performant enough for our visitors.
3. LLM model performance continues to improve, and GPU performance continues to increase, so that even if it's something only someone with multiple video cards can run today, it'll be available to the standard consumer

As the field is emerging rapidly, I wouldn't be surprised if a model/setup emerges that can drive 5000+ guests within the next couple of years on standard hardware. You're right that it's probably only good as a toy example today and can't drive a real park, but in the few months it'll probably take to slowly build the bindings for the agents, the rapid progress in the field might provide a solution for us.

**Reply by ocalhoun6** [reply]
As entertaining as it would be to drown guests who are actually somewhat intelligent, this really doesn't seem like a realistic way to go.

Even if most people have vastly improved hardware and the software gets better ... we still want people to be able to play on low-end hardware. Maybe on hardware that doesn't have a dedicated GPU at all.

**Reply by dvasdekis** [reply]
@ocalhoun6 Thanks for replying!

You are right in that it should not replace the default behaviour. The only requirements would be to amend the default agent (visitor) construct with some more bindings, allowing a plugin to modify them. I believe this would have a trivial performance impact. However, if a user was to enable the plugin, you are also right in that only higher-spec PCs with dedicated graphics cards (for now) could run it. 

For this reason, it would still accomplish your aims, in that it would still allow "people to be able to play on low-end hardware", but without the AI plugin being active (and therefore having the visitors behave as they do today). Unless I'm mistaken?

**Reply by dvasdekis** [reply]
As motivation and food for thought, it's worth reading this paper: https://mobile.twitter.com/emollick/status/1645499660402925576
Consider the possibility of your park users getting truly stressed out and vomiting after they see someone drown, or any one of a multitude of unplanned and comedic/dramatic human-like actions occurring.

**Reply by James103** [reply]
> As motivation and food for thought, it's worth reading this paper

For those who don't want to go through Twitter, the exact paper is at this link: https://arxiv.org/abs/2304.03442 (click "Download PDF" to view the paper).

Regarding the AI-guest interaction, would the more sophisticated guest AI have an influence on the guest's happiness stat and thought list?

**Reply by dvasdekis** [reply]
> Regarding the AI-guest interaction, would the more sophisticated guest AI have an influence on the guest's happiness stat and thought list?

I think it would be more powerful if it did - e.g. those who witness a drowning or an accident in their immediate vicinity should have a much greater hit to happiness than someone who is merely at the park when it occurs.

But still, I think this is getting ahead of the intent of this request - baby steps first :) I guess, I'm requesting a way to read more details about visitor state as the first step - what can they see, what's their happiness, etc. I can then go away and wire those up as model inputs, and at least provide a "debug" output of what the LLM would do, if it was able to. On this basis, I'd then be able to request the ability to edit specific aspects of the visitor object.

**Reply by ZehMatt** [reply]
As interesting as all this sounds this will come down the same road block as everything else with guests/peeps/staff, the code is not in the best state and any attempts to further hack something into it is quite unrealistic to ever happen. I think we should start out with a basic behavior trees where we have it all defined as data first, decoupling behavior from the code also makes it much easier to extend and hack it and we probably don't need neuronal networks to have quite interesting behaviors.

**Reply by dvasdekis** [reply]
Thanks @ZehMatt - could you give an example of where you'd like the code to be refactored, and some example from another project or guide as to what you're looking for? I could give it a go.
